kytos/core/link.py

```diff
@@ -121,65 +123,67 @@ def id(self):  # pylint: disable=invalid-name
+        tag_iterator = Interface.range_intersection(
+            self.endpoint_a.available_tags[tag_type],
+            self.endpoint_b.available_tags[tag_type],
+        )
+        available_tags = list(tag_iterator)
+        return available_tags
 
-    def is_tag_available(self, tag):
+    def is_tag_available(self, tag: int, tag_type: str = 'vlan'):
         """Check if a tag is available."""
-        return (self.endpoint_a.is_tag_available(tag) and
-                self.endpoint_b.is_tag_available(tag))
-
-    def get_next_available_tag(self):
+        return (self.endpoint_a.is_tag_available(tag, tag_type) and
+                self.endpoint_b.is_tag_available(tag, tag_type))
+
+    def get_next_available_tag(
+        self,
+        controller,
+        link_id,
+        tag_type: str = 'vlan'
+    ) -> int:
         """Return the next available tag if exists."""
-        with self._get_available_vlans_lock:
-            # Copy the available tags because in case of error
-            # we will remove and add elements to the available_tags
-            available_tags_a = self.endpoint_a.available_tags.copy()
-            available_tags_b = self.endpoint_b.available_tags.copy()
-
-            for tag in available_tags_a:
-                # Tag does not exist in endpoint B. Try another tag.
-                if tag not in available_tags_b:
-                    continue
-
-                # Tag already in use. Try another tag.
-                if not self.endpoint_a.use_tag(tag):
-                    continue
-
-                # Tag already in use in B. Mark the tag as available again.
-                if not self.endpoint_b.use_tag(tag):
-                    self.endpoint_a.make_tag_available(tag)
-                    continue
-
-                # Tag used successfully by both endpoints. Returning.
-                return tag
-
-            raise KytosNoTagAvailableError(self)
-
-    def make_tag_available(self, tag):
+        with self._get_available_vlans_lock[link_id]:
+            with self.endpoint_a._tag_lock:
+                with self.endpoint_b._tag_lock:
+                    ava_tags_a = self.endpoint_a.available_tags[tag_type]
+                    ava_tags_b = self.endpoint_b.available_tags[tag_type]
+                    tags = Interface.range_intersection(ava_tags_a,
+                                                        ava_tags_b)
+                    try:
+                        tag, _ = next(tags)
+                        self.endpoint_a.use_tags(
+                            controller, [tag, tag], use_lock=False
+                        )
+                        self.endpoint_b.use_tags(
+                            controller, [tag, tag], use_lock=False
+                        )
+                        return tag
+                    except StopIteration:
+                        raise KytosNoTagAvailableError(self)
+
+    def make_tags_available(
+        self,
+        controller,
+        tag: int,
```

Comment:
I believe it'd be a nicer API if `tag` could be either a list[int] or int, the underlying associated interface dependency already takes a list of int. Although we know that currently for links the typical usage is just to make 1 available tag, but since the API is evolving let's keep in a way where it's also compatible and maximize the underlying one, if it also can take here a list[int] (in the same method), then the name of the method becomes coherent. Let me know what you think.

Reply:
Sure thing, although I thought this change was going to be made with `epic_mef_eline_vlan_range` even though a lot of the current changes were made with that epic in mind too. Changing this.

Reply:
Made this changed, `make_tags_available` and `use_tags` accept `Union[int, list[int]]` instead of Link method changing the value to suit code.

Reply:
Nice. The method ended up very ergonomic to use.